Model/AsyncManagement/Charge.php

```diff
@@ -259,11 +259,58 @@ public function decline($order, $chargeId, $detail)
      */
     public function cancel($order, $detail)
     {
-        if (!$order->isCanceled()) {
+        if (!$order->isCanceled() && $order->canCancel()) {
             $order->addStatusHistoryComment($detail['reasonCode'] . ' - ' . $detail['reasonDescription']);
             $order->cancel();
             $order->save();
             $this->asyncLogger->info('Canceled Order #' . $order->getIncrementId());
+        } else {
+            // Get the reason why the order can't be canceled
+            if (!$order->canVoidPayment()) {
+                $order->addStatusHistoryComment("Order cancellation attempted, but failed - payment could not be voided.");
+                $this->asyncLogger->info("Order cancellation attempted, but failed - payment could not be voided, Order #" . $order->getIncrementId());
+            }
+            if ($order->canUnhold()) {
+                $order->addStatusHistoryComment("Order cancellation attempted, but failed - order is on hold.");
+                $this->asyncLogger->info("Order cancellation attempted, but failed - order is on hold, Order #" . $order->getIncrementId());
+            }
+            if (!$order->canReviewPayment() && $order->canFetchPaymentReviewUpdate()) {
```

Comment:
Should the fetch payment method be checked as a negative?

Reply:
I mapped this functionality from the core with notes, so the merchant can understand the reason why it hasn’t been canceled. This way, when Amazon intends to cancel an order but fails, we can understand why.

https://github.com/magento/magento2/blob/98f028aba4b7349646fadd86221e4660de5d4fdc/app/code/Magento/Sales/Model/Order.php#L641